Sources/Subprocess/Configuration.swift

```diff
@@ -589,36 +589,45 @@ internal enum StringOrRawBytes: Sendable, Hashable {
     }
 }
 
-/// A wrapped `FileDescriptor` and whether it should be closed
-/// automatically when done.
-internal struct TrackedFileDescriptor: ~Copyable {
-    internal var closeWhenDone: Bool
-    internal let fileDescriptor: FileDescriptor
-
-    internal init(
-        _ fileDescriptor: FileDescriptor,
-        closeWhenDone: Bool
-    ) {
-        self.fileDescriptor = fileDescriptor
-        self.closeWhenDone = closeWhenDone
-    }
-
-    #if os(Windows)
-    consuming func consumeDiskIO() -> FileDescriptor {
-        let result = self.fileDescriptor
-        // Transfer the ownership out and therefor
-        // don't perform close on deinit
-        self.closeWhenDone = false
-        return result
-    }
+internal enum _CloseTarget {
+    #if canImport(WinSDK)
+    case handle(HANDLE)
     #endif
+    case fileDescriptor(FileDescriptor)
+    case dispatchIO(DispatchIO)
+}
 
-    internal mutating func safelyClose() throws {
-        guard self.closeWhenDone else {
-            return
+internal func _safelyClose(_ target: _CloseTarget) throws {
+    switch target {
+    #if canImport(WinSDK)
+    case .handle(let handle):
+        /// Windows does not provide a “deregistration” API (the reverse of
+        /// `CreateIoCompletionPort`) for handles and it it reuses HANDLE
+        /// values once they are closed. Since we rely on the handle value
+        /// as the completion key for `CreateIoCompletionPort`, we should
+        /// remove the registration when the handle is closed to allow
+        /// new registration to proceed if the handle is reused.
+        AsyncIO.shared.removeRegistration(for: handle)
```

Comment:
is this really a synchronous function? CC @al45tair , will this synchronously & without blocking remove a registration?

Reply:
On Windows, there's no way to remove a registration anyway, so this is just manipulating an internal data structure.